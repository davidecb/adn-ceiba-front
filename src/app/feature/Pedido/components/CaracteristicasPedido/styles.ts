import styled from 'styled-components';

export const CaracteristicasPedidoContainer = styled.div`
  width: 20%;
  padding: 15px;
  position: fixed;
  left: 80%;
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  justify-content: center;
  @media (max-width: 900px) {
    position: static;
    left: 0;
    width: 100%;
  }
`;

export const FormularioContainer = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  z-index: 100;
  background: rgba(0, 0, 0, 0.95);
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
`;
