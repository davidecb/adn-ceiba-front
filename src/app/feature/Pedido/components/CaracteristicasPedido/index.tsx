import * as PropTypes from 'prop-types';
import * as React from 'react';
import { CaracteristicasPedidoContainer, FormularioContainer } from './styles';
import { BtnSolicitarPedido } from '../BtnSolicitarPedido';
import { FormSolicitarPedido } from '../FormSolicitarPedido';
import { Pedido } from '../../models/Pedido';

export interface CaracteristicasPedidoProps {
  pedido: Pedido;
  onClickFinalizarPedido: (pedido: Pedido) => void;
}

export const CaracteristicasPedido: React.FC<CaracteristicasPedidoProps> = ({
  pedido,
  onClickFinalizarPedido,
}) => {

  const [finalizarPedidoClicked, setFinalizarPedidoClicked] = React.useState(false);
  return (
    <CaracteristicasPedidoContainer>
      <div><b>numero de pedido:</b> {pedido.numeroPedido}</div>
      <div><b>cantidad productos:</b> {pedido.productosSolicitados.length}</div>
      <div><b>costo total:</b> ${pedido.costo}</div>
      <div><b>tiempo total:</b> {pedido.tiempo} min.</div>
      <BtnSolicitarPedido pedido={pedido} onSolicitar={() => {
        setFinalizarPedidoClicked(true);
      }} />
      {finalizarPedidoClicked && 
        <FormularioContainer>
          <FormSolicitarPedido 
            pedido={pedido} 
            onSubmit={(pedido) => {
              onClickFinalizarPedido(pedido); 
            }}
            cerrarFormulario={() => {
              setFinalizarPedidoClicked(false);
            }}
          />
        </FormularioContainer>      
      }
    </CaracteristicasPedidoContainer>    
  );
};

CaracteristicasPedido.propTypes = {
  pedido: PropTypes.any.isRequired,
  onClickFinalizarPedido: PropTypes.func.isRequired,
};
