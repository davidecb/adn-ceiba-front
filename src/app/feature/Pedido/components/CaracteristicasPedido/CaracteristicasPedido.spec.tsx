import * as React from 'react';
import { RenderResult, render, fireEvent, wait } from '@testing-library/react';
import { SinonStub, stub } from 'sinon';
import { CaracteristicasPedido } from './../CaracteristicasPedido';
import { setTextEvent } from 'app/shared/utils/test';

describe('CaracteristicasPedido Test', () => {
  let componentWrapper: RenderResult;
  let componentProps: React.ComponentProps<typeof CaracteristicasPedido> & {
    onClickFinalizarPedido: SinonStub;
  };

  beforeEach(() => {
    componentProps = {
      pedido: {
        id: 1,
        numeroPedido: '123456789',
        direccion: '',
        cliente: '',
        estado: '',
        costo: 10000,
        tiempo: 30,
        productosSolicitados:[]
      },
      onClickFinalizarPedido: stub(),
    };
    componentWrapper = render(<CaracteristicasPedido {...componentProps} />);
  });

  it('should match snapshot', () => {
    expect(componentWrapper.container).toMatchSnapshot();
  });

  it('should show formSolicitarPedido on FinalizarPedido click', async () => {
    const elem = componentWrapper.container;
    const solicitarButton = elem.querySelectorAll('button')[0];
    
    await wait(() => {
      solicitarButton && fireEvent.click(solicitarButton);
    });

    const divs = elem.querySelectorAll('div');
    const forms = elem.querySelectorAll('form');
    const h2s = elem.querySelectorAll('h2');

    expect(divs.length).toBe(7);
    expect(forms.length).toBe(1);
    expect(h2s.length).toBe(1);
    expect(h2s[0].textContent).toBe('Solicitar Pedido');
  });

  it('should close formSolicitarProducto on cancelar button click', async () => {
    const elem = componentWrapper.container;
    const solicitarButton = elem.querySelectorAll('button')[0];
    
    await wait(() => {
      solicitarButton && fireEvent.click(solicitarButton);
    });

    const cancelarButton = elem.querySelectorAll('span')[1];
    expect(cancelarButton.textContent).toBe('❌');

    await wait(() => {
      cancelarButton && fireEvent.click(cancelarButton);
    });

    const divs = elem.querySelectorAll('div');
    const forms = elem.querySelectorAll('form');

    expect(divs.length).toBe(5);
    expect(forms.length).toBe(0);
  });
  
  it('should submit and close the form when click on agregar al carrito button', async () => {
    const elem = componentWrapper.container;
 
    const direccion = elem.querySelector('input[name="direccion"]');
    const cliente = elem.querySelector('input[name="cliente"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      direccion && fireEvent.change(direccion, setTextEvent('direccion', 'test street'));
    });
    await wait(() => {
      cliente && fireEvent.change(cliente, setTextEvent('cliente', 'test customer'));
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });

    const divs = elem.querySelectorAll('div');
    const forms = elem.querySelectorAll('form');

    expect(divs.length).toBe(5);
    expect(forms.length).toBe(0);
  });
});

export {};
