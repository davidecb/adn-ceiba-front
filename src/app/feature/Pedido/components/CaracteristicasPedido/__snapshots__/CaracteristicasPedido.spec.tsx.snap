// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaracteristicasPedido Test should match snapshot 1`] = `
<div>
  <div
    class="sc-AxjAm dlYdrF"
  >
    <div>
      <b>
        numero de pedido:
      </b>
       
      123456789
    </div>
    <div>
      <b>
        cantidad productos:
      </b>
       
      0
    </div>
    <div>
      <b>
        costo total:
      </b>
       $
      10000
    </div>
    <div>
      <b>
        tiempo total:
      </b>
       
      30
       min.
    </div>
    <button
      class="sc-AxiKw leNlab"
    >
      <span>
        Realizar Pedido
      </span>
    </button>
  </div>
</div>
`;
